File: csharp/845.longest-mountain-in-array.cs
/*
 * @lc app=leetcode id=845 lang=csharp
 *
 * [845] Longest Mountain in Array
 */

using System;

// @lc code=start
public class Solution {
    public int LongestMountain(int[] A) {
        int ret = 0, left = 0, peak = 0;
        for (int i = 1; i < A.Length; ++i) {
            if (A[i] > A[i - 1]) {
                if (peak < i - 1) {
                    left = i - 1;
                }
                peak = i;
            } else if (A[i] == A[i - 1]) {
                left = i;
                peak = i;
            } else {
                if (left != peak) {
                    ret = Math.Max(i - left + 1, ret);
                } else {
                    left = i;
                    peak = i;
                }
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/845.longest-mountain-in-array.java
/*
 * @lc app=leetcode id=845 lang=java
 *
 * [845] Longest Mountain in Array
 */

// @lc code=start
class Solution {
    public int longestMountain(int[] arr) {
        int ret = 0, l = 0, p = 0, n = arr.length;
        for (int i = 1; i < n; ++i) {
            if (arr[i] > arr[i - 1]) {
                if (p < i - 1) {
                    l = i - 1;
                }
                p = i;
            } else if (arr[i] < arr[i - 1] && p != l) {
                ret = Math.max(i - l + 1, ret);
            } else {
                l = i;
                p = i;
            }
        }
        return ret;
    }
}
// @lc code=end

